add retry classifier customization RFC (#3018)
awslabs/smithy-rs: design/src/rfcs/rfc0037_retry_classifier_customization.md

# RFC: User-configurable retry classification

> Status: RFC
>
> Applies to: client

For a summarized list of proposed changes, see the [Changes Checklist](#changes-checklist) section.

> This RFC defines the user experience and implementation of user-configurable retry classification. Custom retry classifiers enable users to change what errors are retried while still allowing them to rely on defaults set by SDK authors when desired.

Reviewer: ```suggestion
This RFC defines the user experience and implementation of user-configurable retry classification. Custom retry classifiers enable users to change what responses are retried while still allowing them to rely on defaults set by SDK authors when desired.
```

## Terminology

- **Smithy Service**: An HTTP service, whose API is modeled with the [Smithy IDL](https://www.smithy.io).
- **Smithy Client**: An HTTP client generated by smithy-rs from a `.smithy` model file.
- **AWS SDK**: A **smithy client** that's specifically configured to work with an AWS service.

> - **Operation**: A modeled interaction with a service, defining the proper input and expected output types, as well as important metadata related to request construction. "Sending" an operation implies sending one or more HTTP requests to **smithy service**, and then receiving an output or error in response.

Reviewer: ```suggestion
- **Operation**: A modeled interaction with a service, defining the proper input and expected output shapes, as well as important metadata related to request construction. "Sending" an operation implies sending one or more HTTP requests to a **Smithy service**, and then receiving an output or error in response.
```

- **Orchestrator**: The client code which manages the request/response pipeline. The orchestrator is responsible for: 
	- Constructing, serializing, and sending requests.
	- Receiving, deserializing, and (optionally) retrying requests.
	- Running interceptors *(not covered in this RFC)* and handling errors.

> - **Runtime Component**: A part of the orchestrator responsible for a specific function. Runtime components are *always* required and may depend on specific configuration. Examples include the endpoint resolver, retry strategy, and request signer.

Reviewer: ```suggestion
- **Runtime Component**: A part of the orchestrator responsible for a specific function. Runtime components are used by the orchestrator itself, may depend on specific configuration, and must not be changed by interceptors. Examples include the endpoint resolver, retry strategy, and request signer.
```
The `http_client`/`http_connector` and `sleep_impl` components are actually optional. The HTTP piece isn't used for presigned requests, and the async sleep isn't used in WASM targets.

> - **Runtime Plugin**: Code responsible for setting and **runtime components** and related configuration. Runtime plugins defined by codegen are responsible for setting default configuration and altering the behavior of **smithy clients** including the **AWS SDKs**.

Reviewer: ```suggestion
- **Runtime Plugin**: Code responsible for setting and **runtime components** and related configuration. Runtime plugins defined by codegen are responsible for setting default configuration and altering the behavior of **Smithy clients** including the **AWS SDKs**.
```

- **Retry Strategy**: The process by which the orchestrator determines when and how to retry failed requests. Only one retry strategy may be set. The retry strategy depends upon the **retry classifier** to interpret responses and determine if they are retryable.

> - **Retry Classifier**: Code responsible for introspecting the request/response pipeline's and determining if a retry is necessary. Multiple retry classifiers may be combined into a **retry classifier chain**.

Reviewer: ```suggestion
- **Retry Classifier**: Code responsible for introspecting the request/response pipelines and determining if a retry is necessary. Multiple retry classifiers may be combined into a **retry classifier chain**.
```

> - **Retry Classifier Chain**: Requests can fail for any number of reasons, but retry classifiers are specific, meaning that they target a narrow range of possibilities. Rather than define a single retry classifier to handle all possible outcomes, multiple classifiers are chained together. Each classifier is run in turn, and the chain ends when a classifier decides a response is retryable, or a classifier may determine that a response must not be retried, or after all classifiers have run and haven't decided a retry is necessary.

Reviewer: I think the retry classifier pieces should be defined as part of the RFC rather than the terminology for making sense of the RFC.

```rust
#[tokio::main]
async fn main() -> Result<(), aws_sdk_s3::Error> {
    let sdk_config = aws_config::load_from_env().await;
    let client = aws_sdk_s3::Client::new(&sdk_config);

    let res = client
	    .list_buckets()
        .customize()
        .await
        .unwrap()
        .config_override(   
            aws_sdk_s3::Config::builder()
	            .retry_classifier(CustomRetryClassifier)
        )
        .send()
        .await?;

	println!("your buckets: {res:?}");

    Ok(())
}
```

## How to actually implement this RFC

In order to implement this feature, we must:
- Update the current retry classification system so that individual classifiers as well as collections of classifiers can be easily composed together.
- Create two new configuration mechanisms for users that allow them to customize retry classification at the service level and at the operation level.
- Update retry classifiers so that they may 'short-circuit' the chain, ending retry classification immediately.

### The `RetryClassifier` trait

```rust
/// The result of running a [`ClassifyRetry`] on a [`InterceptorContext`].
#[non_exhaustive]
#[derive(Clone, Eq, PartialEq, Debug)]

> pub enum RetryClassifierResult {
>     /// "A retryable error was received. This is what kind of error it was,
>     /// in case that's important."
>     Error(ErrorKind),
>     /// "The server told us to wait this long before retrying the request."
>     Explicit(Duration),
>     /// "This response should not be retried."
>     DontRetry,
> }

Reviewer: What do you think about calling this `RetryAction`? I think we could also revisit some of the other names now that the design has drifted quite a ways away from what it was originally.

What is the `ErrorKind` used for again? Should `RetryAfter` have `ErrorKind`?

```suggestion
pub enum RetryAction {
    /// "A retryable error was received. This is what kind of error it was,
    /// in case that's important."
    Retry(ErrorKind),
    /// "The server told us to wait this long before retrying the request."
    RetryAfter(Duration),
    /// "This response should not be retried."
    NoRetry,
}
```

/// Classifies what kind of retry is needed for a given [`InterceptorContext`].  
pub trait ClassifyRetry: Send + Sync + fmt::Debug {
    /// Run this classifier on the [`InterceptorContext`] to determine if the previous request
    /// should be retried. Returns a [`RetryClassifierResult`].  
    fn classify_retry(
        &self,
        ctx: &InterceptorContext,

>         result_of_preceding_classifier: Option<RetryClassifierResult>,

Reviewer: Could be `preceding_action` or `preceding_class`

>     ) -> Option<RetryClassifierResult>;

Reviewer: What does returning `None` mean? What happens if all the classifiers return `None`?

Should retry classifiers by fallible?

Author: I'll clarify.

Author: `None` means that a classifier had nothing to say. For example, this would happen when a classifier that relies on successful response deser encounters a deser failure. If all return `None`, then no retry is necessary

    /// The name of this retry classifier.
    ///
    /// Used for debugging purposes
    fn name(&self) -> &'static str;

    /// The priority of this retry classifier. Classifiers with a higher priority will run before
    /// classifiers with a lower priority. Classifiers with equal priorities make no guarantees
    /// about which will run first.

>     fn priority(&self) -> RetryClassifierPriority {

Reviewer: This priority type should be declared somewhere in the RFC

        RetryClassifierPriority::default()
    }
}
```

### Chaining retry classifiers

> Multiple retry classifiers are chained by wrapping classifiers inside one another. When classifiers must be wrapped in a specific order, use a specific type for the inner classifier. When classifiers must be composable in any order, use a referenced trait object. This approach is demonstrated in the following example code:

Reviewer: Is this still accurate, or is it using priority-based sorting now?

Author: It's no longer accurate. I'll rework this section.

### Setting a retry classifier with a runtime plugin

Wrapping retry classifiers may be set with a runtime plugin. When setting a classifier with this method, the runtime plugin is responsible for extracting any previously-set classifier and wrapping it.

```rust
impl<'a> ExampleRetryClassifier<'a> {
	pub fn new(inner: Option<&'a dyn ClassifyRetry>) -> Self {
		Self { inner }
	}
}

struct ExampleRetryClassifierRuntimePlugin;

impl RuntimePlugin for ExampleRetryClassifierRuntimePlugin {
	fn runtime_components(  
	    &self,  
	    current_components: &RuntimeComponentsBuilder,  
	) -> Cow<'_, RuntimeComponentsBuilder> {
	
	    let rcb = RuntimeComponentsBuilder::new("ExampleRetryClassifierRuntimePlugin")
		    .with_retry_classifier(
			    ExampleRetryClassifier::new(current_components.retry_classifier())
		    );

	    Cow::Owned(rcb)
	}
}
```

By default, newer runtime plugins will override previously-set plugins. This is important to consider when deciding how your classifier will wrap other classifiers.

> 

Reviewer: I think there's a missing section on standard classifiers provided by smithy-rs, and how to reference them or disable them.

## Changes checklist

- [ ] Make retry classifiers composable by runtime plugins.
- [ ] Enable configuration of retry classifiers at the service level.
- [ ] Enable configuration of retry classifiers at the operation level.
- [ ] Replace `RetryReason` with `RetryClassifierResult`.
	- [ ] Add variant for `DontRetry`

> 	- [ ] Add variant for `Continue`

Reviewer: `Continue` is not included above.
